src/testSession.ts

```diff
@@ -0,0 +1,234 @@
+import { TestProject, TestProjectConfig } from './testProject';
+
+export interface TestSessionOptions {
+  /**
+   * Specify a different location for the test session.
+   */
+  sessionDir?: string;
+
+  /**
+   * Define a test project to use for tests.
+   */
+  project?: TestProjectConfig;
+
+  /**
+   * Commands to run as setup for tests.  All must have exitCode == 0 or session
+   * creation will throw.  If an org create command is specified the test session
+   * will have a `orgUsername` property for use in tests.
+   */
+  setupCommands?: string[];
+}
+
+// Executes commands and sets the org username if an org was created.
+// Throws if any commands return a non-zero exitCode.
+const setupCommands = (testSession: TestSession, cmds?: string[]): void => {
+  if (cmds) {
+    const dbug = debug('testkit:setupCommands');
+    for (let cmd of cmds) {
+      // if it's an org:create without --json, append it anyway so we
+      // can get the org username and keep track of it.
+      if (cmd.includes('org:create')) {
+        if (testSession.orgUsername) {
+          return;
+        }
+        if (!cmd.includes('--json')) {
+          cmd += ' --json';
+        }
+      }
+      const rv = shell.exec(cmd, { silent: true });
+      if (rv.code !== 0) {
+        const io = cmd.includes('--json') ? rv.stdout : rv.stderr;
+        throw Error(`Setup command ${cmd} failed due to: ${io}`);
+      }
+
+      // keep track of org creates
+      if (cmd.includes('org:create')) {
+        try {
+          dbug(`Saving username from ${cmd}`);
+          const jsonOutput = parseJson(rv.stdout);
+          testSession.orgUsername = getString(jsonOutput, 'result.username') || undefined;
+        } catch (err: unknown) {
+          dbug(`TestSession created an org but failed JSON parsing due to:\n${(err as Error).message}`);
+        }
+      }
+    }
+  }
+};
+
+/**
+ * Represents a test session, which is a unique location for non-unit test (nut)
+ * artifacts such as a project and a mocked home dir.  It also provides easy
+ * access to org usernames created by setup commands, cwd stubbing, and a way to
+ * zip up the test session.
+ *
+ * Create a TestSession instance with: `const testSession = TestSession.create(options)`
+ *
+ * Fine-grained control over certain test session details are provided by these
+ * environment variables:
+ *   TESTKIT_SESSION_DIR = Overrides the default directory for the test session
+ *   TESTKIT_HOMEDIR = path to a home directory that the tests will use as a stub of os.homedir
+ *   TESTKIT_ORG_USERNAME = an org username to use for test commands. tests will use this org rather than creating new orgs.
+ *   TESTKIT_PROJECT_DIR = a SFDX project to use for testing. the tests will use this project directly.
+ *   TESTKIT_SAVE_ARTIFACTS = prevents a test session from deleting orgs, projects, and test sessions.
+ *   TESTKIT_ZIP_SESSION = zips the session dir when this is true
+ */
+export class TestSession {
```

Comment:
TestSession is only tracking a single orgUsername, yet comments indicate otherwise. Is that work still pending?

Reply:
At this point it should all be singular so I'll update all comments to refer to a single org created in setup.  This is one decision that could easily change, but using the 80/20 rule we thought that most plugins would use 1 scratch org.  If multiple are needed then the plugin authors can handle that themselves, and may want to use other methods for scratch org creation such as creating multiple orgs in parallel in a global fixture.